cli/Sources/TuistServer/Client/ServerAuthenticationController.swift

```diff
@@ -74,92 +88,175 @@ public struct ServerAuthenticationController: ServerAuthenticationControlling {
         #endif
     }
 
+    public func refreshToken(serverURL: URL, inSubprocess: Bool) async throws {
+        try await authenticationTokenRefreshingIfNeeded(
+            serverURL: serverURL,
+            forceRefresh: true,
+            inSubprocess: inSubprocess
+        )
+    }
+
     public func refreshToken(serverURL: URL) async throws {
         try await authenticationTokenRefreshingIfNeeded(
-            serverURL: serverURL, forceRefresh: true
+            serverURL: serverURL,
+            forceRefresh: true,
+            inSubprocess: true
         )
     }
 
     @discardableResult private func authenticationTokenRefreshingIfNeeded(
         serverURL: URL,
-        forceRefresh: Bool
+        forceRefresh: Bool,
+        inSubprocess: Bool
     ) async throws
         -> AuthenticationToken?
     {
         #if canImport(TuistSupport)
             Logger.current.debug("Refreshing authentication token for \(serverURL) if needed")
         #endif
-        return
-            try await CachedValueStore.current
-                .getValue(key: "token_\(serverURL.absoluteString)") {
-                    () -> (value: AuthenticationToken, expiresAt: Date?)? in
-                    guard let token = try await fetchTokenFromStore(serverURL: serverURL) else {
-                        return nil
-                    }
+        switch (try await tokenStatus(serverURL: serverURL, forceRefresh: forceRefresh), inSubprocess) {
+        case let (.valid(token), _):
+            return token
+        case (.absent, _), (_, false):
+            return try await executeRefresh(serverURL: serverURL, forceRefresh: forceRefresh)?.value
+        case (.expired, true):
+            let lockfilePath = lockFilePath(serverURL: serverURL)
+
+            if !(try await fileSystem.exists(lockfilePath)) {
+                try await fileSystem.touch(lockfilePath)
+
+                let process = Process()
+                process.environment = ProcessInfo.processInfo.environment
+                process.launchPath = Environment.current.currentExecutablePath()?.pathString ?? ""
+                process.arguments = [
+                    "auth",
+                    "refresh-token",
+                    "--server-url",
+                    serverURL.absoluteString,
+                    "--lockfile-path",
+                    lockfilePath.pathString,
+                ]
+                process.unbind(.isIndeterminate)
+                try process.run()
+            }
+            try await Task.sleep(nanoseconds: 200_000_000) // 200 miliseconds
+            return try await authenticationTokenRefreshingIfNeeded(
```

Comment:
I need to add a countdown here so that we don't loop indefinitely if there's a bug in the refresh logic.